server/src/test/java/org/elasticsearch/indices/recovery/RecoverySettingsTests.java

```diff
@@ -89,4 +103,199 @@ public void testMaxConcurrentSnapshotFileDownloadsPerNodeIsValidated() {
             )
         );
     }
+
+    public void testDefaultMaxBytesPerSecOnNonDataNode() {
+        assertThat(
+            "Non-data nodes have a default 40mb rate limit",
+            nodeRecoverySettings().withRole(randomFrom("master", "ingest", "ml")).withRandomMemory().build().getMaxBytesPerSec(),
+            equalTo(DEFAULT_MAX_BYTES_PER_SEC)
+        );
+    }
+
+    public void testMaxBytesPerSecOnNonDataNodeWithIndicesRecoveryMaxBytesPerSec() {
+        final ByteSizeValue random = randomByteSizeValue();
+        assertThat(
+            "Non-data nodes should use the defined rate limit when set",
+            nodeRecoverySettings().withRole(randomFrom("master", "ingest", "ml"))
+                .withIndicesRecoveryMaxBytesPerSec(random)
+                .withRandomMemory()
+                .build()
+                .getMaxBytesPerSec(),
+            equalTo(random)
+        );
+    }
+
+    public void testDefaultMaxBytesPerSecOnDataNode() {
+        assertThat(
+            "Data nodes that are not dedicated to cold/frozen have a default 40mb rate limit",
+            nodeRecoverySettings().withRole(randomFrom("data", "data_hot", "data_warm", "data_content"))
+                .withRandomMemory()
+                .build()
+                .getMaxBytesPerSec(),
+            equalTo(DEFAULT_MAX_BYTES_PER_SEC)
+        );
+    }
+
+    public void testMaxBytesPerSecOnDataNodeWithIndicesRecoveryMaxBytesPerSec() {
+        final ByteSizeValue random = randomByteSizeValue();
+        assertThat(
+            "Data nodes that are not dedicated to cold/frozen should use the defined rate limit when set",
+            nodeRecoverySettings().withRole(randomFrom("data", "data_hot", "data_warm", "data_content"))
```

Comment:
Can we randomly add either `data_cold` or `data_frozen` (or both) to the set of roles here?

Reply:
You spotted it :)